### ISSUE TITLE
Updating attribute values not reflected in variation dropdown

### ISSUE THREAD

**Issue by rickgregory** [woocommerce/woocommerce#23111]
**Describe the bug**
Editing the values of an attribute after the first save appears to succeed but the new values are not reflected in the variation dropdown on the single product page

**To Reproduce**
1) Create a fresh WP install with the current version of WC (3.5.7 right now). 
2) Add a variable product. 
3) Add one attribute and give it a value (can be several pipe delimited values)
4) Check the Use as Variation checkbox for this attribute.
5) Save Attribute and publish product. 
6) view single product page. Note the values in the dropdown reflect what you assigned in #3.
7) Edit the product and alter the attribute values in some fashion. Again, Save Attributes and update the product
8) look at the single product page.

Actual: Old attribute values are shown in the dropdown
Expected: New attribute values shown. 

More: If you open variables.php in the plugin includes, find the code that populates the dropdown and edit it like this https://gist.github.com/rickgregory/dcf0563f20031040962ce6add3a886b5 you'll see the raw values... and they are the old values first entered, not the new edited values. 
 
<img width="761" alt="Screen Shot 2019-03-21 at 9 36 42 PM" src="https://user-images.githubusercontent.com/1743922/54800904-7d388a80-4c21-11e9-8888-525960d914ac.png">
<img width="777" alt="Screen Shot 2019-03-21 at 9 36 28 PM" src="https://user-images.githubusercontent.com/1743922/54800905-7d388a80-4c21-11e9-8d5e-fad0a039eed6.png">

**Isolating the problem (mark completed items with an [x]):**
- [x ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
NOTE: THIS IS A TEST SERVER I SPUN UP ON DIGITAL OCEAN, NOT MY PRODUCTION SERVER
`
### WordPress Environment ###

Home URL: http://104.248.79.125
Site URL: http://104.248.79.125
WC Version: 3.5.7
Log Directory Writable: ✔
WP Version: ❌ 5.0.4 - There is a newer version of WordPress available (5.1.1)
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.29 (Ubuntu)
PHP Version: 7.2.10-0ubuntu0.18.04.1
PHP Post Max Size: 16 MB
PHP Time Limit: 60
PHP Max Input Vars: 1000
cURL Version: 7.58.0
OpenSSL/1.1.0g

SUHOSIN Installed: –
MySQL Version: 5.7.25-0ubuntu0.18.04.2
Max Upload Size: 16 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.5.7
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 2.88MB
Database Data Size: 2.04MB
Database Index Size: 0.84MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.02MB + Index: 0.08MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 1.50MB + Index: 0.02MB
wp_postmeta: Data: 0.02MB + Index: 0.03MB
wp_posts: Data: 0.02MB + Index: 0.06MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.02MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB

### Post Type Counts ###

page: 2
post: 2
product: 1
product_variation: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 3.5.7

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

### WC Pages ###

Shop base: ❌ Page not set
Cart: ❌ Page not set
Checkout: ❌ Page not set
My account: ❌ Page not set
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Nineteen
Version: 1.2 – 1.3 is available
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –
`

**Comment by mikejolley** [comment]
@rickgregory I think you may have missed a step. After saving the updated attributes they need assigning the variations on the variations tab. Only attributes which match variations are going to be shown on the page.

**Comment by rickgregory** [comment]
Nope. Just double-checked that. Actually did a short video which I'll post in a while.

The issue is NOT that the variation doesn't show. It does (see screenshot). It's that the data in the dropdown is not updated when you edit the attribute on the backend. In the real version of this, I have a product with the variation assigned. The attribute values were incorrect, so I edited those under Attributes... and the new values DO NOT show. If you look a the addition info tab, they're there. HOWEVER if you use the gist code above the dropdown will show values... the OLD ones.

**Comment by rickgregory** [comment]
OK I found the issue... and it's bad UX. 

Say I create a shirt as a variable product. I create an attribute called Size and assign values of S | M | L. I create variation for each, assign each variation a price, save and publish. The dropdown shows all 3. I then edit the product, adding a new value to the Size attribute, XL. I save that and add a variation for it. Update. Look at the product. No XL in the dropdown. WTF???

Did you spot what I left out? 

I didn't add a price to the XL variation. The UI does NOT warn me about this. It looks like everything is fine... it just fails to show the variation in the dropdown. 

This wouldn't be so bad IF I could add a base or master price that variations all inherited and which I could override if I wanted to. But you cannot do that. When you choose Variable product you have to assign price per variation. 

So this is a UX/UI bug, not a data bug. I imagine that once you get in the flow of adding a price as you add a variation you will not see this... but the user interface should either a) let me set a base price that new variations inherit OR flag the fact, on save, that I'm saving a variation without entering a price.

**Comment by mikejolley** [comment]
Thats it. We're aware the UX is not great for variations. Re-making those is basically on-hold until the UI is updated for Gutenberg/with react. The wc-admin project (https://github.com/woocommerce/woocommerce-admin) is working towards that.

New UIs have already been experimented with on Store for WP.com (https://en.support.wordpress.com/store/products/) where prices work similar to how you'd expect (base with overides per variation).

Currently the field is marked required but as you say, if you miss that your variation is not going to display. As a temp measure I'll see about adding a more prominant warning if variations are missing prices.

**Comment by rickgregory** [comment]
Mike - thanks. One other thing that might be useful in redoing this is for a new variation to inherit the base price. I ran into a situation where someone hadn't defined an attribute as Used In Variations. I did that, then told Woo to create variations. It did... but left the price blank, so of course those variations showed as unavailable. 

For a few SKUs, eh. But this resulted in hundreds of new SKUs so hand editing them was... not useful. 

Let me know if you'd like this written as a new issue vs being here.
